geopandas/tests/test_geodataframe.py

```diff
@@ -1270,6 +1270,64 @@ def test_multiindex_with_geometry_label(self, dtype):
         result = gdf[["geometry"]]
         assert_frame_equal(result, gdf if dtype == "geometry" else pd.DataFrame(gdf))
 
+    def test_multiindex_geometry_colname_2_level(self):
+        # GH1763 https://github.com/geopandas/geopandas/issues/1763
+        crs = "EPSG:4326"
+        df = pd.DataFrame(
+            [[1, 0], [0, 1]], columns=[["location", "location"], ["x", "y"]]
+        )
+        x_col = df["location", "x"]
+        y_col = df["location", "y"]
+
+        gdf = GeoDataFrame(df, crs=crs, geometry=points_from_xy(x_col, y_col))
+        assert gdf.crs == crs
+        assert gdf.geometry.crs == crs
+        assert gdf.geometry.dtype == "geometry"
+        assert gdf._geometry_column_name == ("geometry", "")
+        assert gdf.geometry.name == ("geometry", "")  # or is "geometry" this expected?
```

Comment:
```suggestion
        assert gdf.geometry.name == ("geometry", "")
```

Since I think we established this is expected?

(it might depend on where we use this, as in many places, like in `df[colname]`, just using a single "geometry" string works in pandas, but I can imagine there are some places where it needs to be the exact tuples)

Reply:
Actually, that's not what I see locally:

```
In [44]: crs = "EPSG:4326"
    ...: df = pd.DataFrame(
    ...:     [[1, 0], [0, 1]],
    ...:     columns=[
    ...:         ["foo", "foo"],
    ...:         ["location", "location"],
    ...:         ["x", "y"],
    ...:     ],
    ...: )
    ...: 
    ...: x_col = df["foo", "location", "x"]
    ...: y_col = df["foo", "location", "y"]
    ...: df["geometry"] = gpd.points_from_xy(x_col, y_col)

In [45]: geopandas.GeoDataFrame(df, geometry="geometry").geometry.name
Out[45]: 'geometry'
```

Or did that change with this PR?

Also, it seems that pandas itself gives you the name as a single string when accessing such a column in a pd.DataFrame:

```
In [46]: df["geometry"].name
Out[46]: 'geometry'

In [47]: type(df)
Out[47]: pandas.core.frame.DataFrame

In [48]: df.columns
Out[48]: 
MultiIndex([(     'foo', 'location', 'x'),
            (     'foo', 'location', 'y'),
            ('geometry',         '',  '')],
           )
```

Reply:
So this PR does change that behaviour,
> \# or is "geometry" this expected?

This comment is 18 months old, so I guess the behaviour change is what I was concerned about, I don't remember anymore.
I suppose we should try and align with pandas behaviour, I'll have another look and see if I can figure that out.

Reply:
It does seem pandas does some magic handling of things, because if I extend your example, 
```python
In [4]: df.iloc[:, -1]
Out[4]:
0    POINT (1.00000 0.00000)
1    POINT (0.00000 1.00000)
Name: (geometry, , ), dtype: geometry

In [5]: df.iloc[:, -1].name
Out[5]: ('geometry', '', '')

In [6]: df.columns
Out[6]:
MultiIndex([(     'foo', 'location', 'x'),
            (     'foo', 'location', 'y'),
            ('geometry',         '',  '')],
           )
```
`iloc` doesn't do the same name column name unmangling thing.

Reply:
Having a deeper dig, the issue in the n>=3 case is not actually to do with setting the geometry column, it's to with retrieving the geometry column. Set geometry is perfectly fine with the correct dtype:
```python
In [9]: gpd.GeoDataFrame(df, geometry="geometry").dtypes
Out[9]:
foo       location  x       int64
                    y       int64
geometry                 geometry
dtype: object
```
I will need to re-test on pandas main, but the issue seems to be inside `_getitem_multilevel`
https://github.com/pandas-dev/pandas/blob/5b9f980880776923d65d910935f10c2f4a7e42a3/pandas/core/frame.py#L3846-L3857
the `result=results[""]` call converts the result from a GeoDataFrame with geometry dtype correctly preserved (with an invalid active geometry column, because the columns are renamed as part of the method), to a series with object dtype.

Reply:
I've opened https://github.com/pandas-dev/pandas/pull/51895 which I think would fix this on the pandas side, but that doesn't help so much here, since we'll be waiting a while for this to filter through.

Reply:
Thanks for the pandas PR! While that solves an issue to retain the _dtype_, how does that relate to the discussion above about the name?

Reply:
Good question! That's not clear at all from my comments. With the pandas PR, and the change in this PR removed, the `df.geometry.name` in this context would be strings again (not the tuples that they are with this PR) i.e.
```python
       df = pd.DataFrame(
            [[1, 0], [0, 1]],
            columns=[
                ["foo", "foo"],
                ["location", "location"],
                ["x", "y"],
            ],
        )

        x_col = df["foo", "location", "x"]
        y_col = df["foo", "location", "y"]

        gdf = GeoDataFrame(df, crs=crs, geometry=points_from_xy(x_col, y_col))
        assert gdf.crs == crs
        assert gdf.geometry.crs == crs
        assert gdf.geometry.dtype == "geometry"
        assert gdf._geometry_column_name == "geometry"
        assert gdf.geometry.name == "geometry"
```
holds, rather than the tests as they are written. 

I'm not sure how to resolve this in the geopandas context though, if we wait for the pandas fix that is presumably pandas 2.1 at least, and if we merge this PR as is, df.geometry.name will be tuples which is inconsistent with pandas.

Perhaps it's possible to have another, better workaround on the geopandas side? In theory we could monkey patch / override `GeoDataFrame._getitem_multilevel`, but that is something I would rather avoid.

Reply:
> Perhaps it's possible to have another, better workaround on the geopandas side? In theory we could monkey patch / override `GeoDataFrame._getitem_multilevel`, but that is something I would rather avoid.

Indeed would avoid overriding `_getitem_multilevel` (that's private), but we already do have some custom logic in `__getitem__`, so maybe we could add some kind of check for this issue there?

Reply:
If we were to do this, we either try and intercept before the getitem call, which I think essentially becomes checking for the code path to `_getitem_multilevel` and then providing a patch around that (at which point I wonder if it then would just make sense to override `_getitem_multilevel` conditionally on being an old version of pandas, even if it is private), or after the fact we upcast to geometry again.

I've gone with the latter for what I've just pushed.